src/Components/Server/test/Circuits/RemoteJSDataStreamTest.cs

```diff
@@ -154,10 +155,104 @@ public async void ReceiveData_ProvidedWithOutOfOrderChunk_SimulatesSignalRDiscon
+            {
+                Assert.Equal("Did not receive any data in the alloted time.", ex.Message);
+                Assert.IsType<TimeoutException>(ex);
+                timeoutExceptionRaisedSemaphore.Release();
+            };
+
+            var jsStreamReference = Mock.Of<IJSStreamReference>();
+            var remoteJSDataStream = await RemoteJSDataStream.CreateRemoteJSDataStreamAsync(
+                jsRuntime,
+                jsStreamReference,
+                totalLength: 15,
+                maximumIncomingBytes: 10_000,
+                jsInteropDefaultCallTimeout: TimeSpan.FromSeconds(10), // Note we're using a 10 second timeout for this test
+                pauseIncomingBytesThreshold: 50,
+                resumeIncomingBytesThreshold: 25,
+                cancellationToken: CancellationToken.None);
+            var streamId = GetStreamId(remoteJSDataStream, jsRuntime);
+            var chunk = new byte[] { 3, 5, 7 };
+
+            // Act & Assert 1
+            // Trigger timeout and ensure unhandled exception raised to crush circuit
+            remoteJSDataStream.InvalidateLastDataReceivedTimeForTimeout();
+            await timeoutExceptionRaisedSemaphore.WaitAsync();
+
+            // Act & Assert 2
+            // Confirm exception also raised on pipe reader
+            using var mem = new MemoryStream();
+            var ex = await Assert.ThrowsAsync<TimeoutException>(async () => await remoteJSDataStream.CopyToAsync(mem));
+            Assert.Equal("Did not receive any data in the alloted time.", ex.Message);
+
+            // Act & Assert 3
+            // Ensures stream is disposed after the timeout and any additional chunks aren't accepted
+            var success = await RemoteJSDataStream.ReceiveData(jsRuntime, streamId, chunkId: 0, chunk, error: null);
+            Assert.False(success);
+        }
+
+        [Fact]
+        public async void ReceiveData_ReceivesDataThenTimesout_StreamDisposed()
+        {
+            // Arrange
+            var jsRuntime = new TestRemoteJSRuntime(Options.Create(new CircuitOptions()), Options.Create(new HubOptions()), Mock.Of<ILogger<RemoteJSRuntime>>());
+            var timeoutExceptionRaisedSemaphore = new SemaphoreSlim(initialCount: 0, maxCount: 1);
+            jsRuntime.UnhandledException += (_, ex) =>
+            {
+                Assert.Equal("Did not receive any data in the alloted time.", ex.Message);
+                Assert.IsType<TimeoutException>(ex);
+                timeoutExceptionRaisedSemaphore.Release();
+            };
+
+            var jsStreamReference = Mock.Of<IJSStreamReference>();
+            var remoteJSDataStream = await RemoteJSDataStream.CreateRemoteJSDataStreamAsync(
+                jsRuntime,
+                jsStreamReference,
+                totalLength: 15,
+                maximumIncomingBytes: 10_000,
+                jsInteropDefaultCallTimeout: TimeSpan.FromSeconds(30), // Note we're using a 30 second timeout for this test
```

Comment:
Generally we try to make tests as quick as possible, we do this by
a) not using real timers (mock timers so we control when they tick and how much time they think has passed)
b) avoiding timers if possible (presumably this test is actually testing a timeout so b. doesn't apply)
c) using much lower values for timeouts if a. or b. can't be done

Additionally, relying on a real timeout creates potential flakiness because what if the machine is really slow and the test doesn't do the work you expected before the timeout occurs?

Couple examples:
https://github.com/dotnet/aspnetcore/blob/cee22edd8c27c1ac767a06005d19fd973ab7ee22/src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs#L26-L35
https://github.com/dotnet/aspnetcore/blob/65276b0476c4a004e2dff8c205230a59deb6c627/src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs#L2690-L2708

Last thing, when using `await` you should consider putting [`.DefaultTimeout()`](https://github.com/dotnet/aspnetcore/blob/main/src/Shared/TaskExtensions.cs) on the calls so that any hangs will fail the test (and tell you where they hung) and not timeout the test run instead.
Example: https://github.com/dotnet/aspnetcore/blob/65276b0476c4a004e2dff8c205230a59deb6c627/src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs#L4453-L4461

Sorry if this ended up looking like a bit of a rant, but I really want our tests to be reliable.

Reply:
Thanks @BrennanConroy appreciate the feedback. I went the route of using a `TaskCompletionSource` to match the pattern used in [other `src/Components/tests` tests](https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Components/Components/test/ComponentBaseTest.cs#L186-L236) to keep consistency. I also significantly reduced the timeout values to avoid long running tests, and the `TaskCompletionSource` approach should eliminate flakiness as we'll be able to await the timeout. I've also updated the tests to leverage `DefaultTimeout()`.

https://github.com/dotnet/aspnetcore/pull/33916/commits/1cd162f6eed4f7249081ad9f877ce83456dd35c3